ydb_sqlalchemy/dbapi/connection.py

```diff
@@ -1,24 +1,38 @@
 import posixpath
+from typing import Optional
 
 import ydb
 from .cursor import Cursor
-from .errors import InterfaceError, ProgrammingError, DatabaseError
+from .errors import InterfaceError, ProgrammingError, DatabaseError, InternalError, NotSupportedError
+
+
+class IsolationLevel:
+    SERIALIZABLE = "SERIALIZABLE"
+    ONLINE_READONLY = "ONLINE READONLY"
+    ONLINE_READONLY_INCONSISTENT = "ONLINE READONLY INCONSISTENT"
+    STALE_READONLY = "STALE READONLY"
+    SNAPSHOT_READONLY = "SNAPSHOT READONLY"
+    AUTOCOMMIT = "AUTOCOMMIT"
 
 
 class Connection:
     def __init__(self, endpoint=None, host=None, port=None, database=None, **conn_kwargs):
         self.endpoint = endpoint or f"grpc://{host}:{port}"
         self.database = database
-        self.driver = self._create_driver(self.endpoint, self.database, **conn_kwargs)
-        self.pool = ydb.SessionPool(self.driver)
+        self.table_client_settings = self._get_table_client_settings()
+        self.driver = self._create_driver(**conn_kwargs)
```

Comment:
I suggest about driver and pool session should be one per engine, not one per connection if possible.

1. Every driver instance and session pool has own background processes: for cluster rediscovery, ping sessions in pool, etc. 
Every additional driver instance has additional overhead. Every driver initialization need few seconds.

2. Session is not very stable - session may be closed often and re-create session it is normal process during application work (instead of reconnect to postgres, where connection fail is trouble situation).

As target state it should be one session pool per sqlalchemy engine and reuse it in connections.
If sqlalchemy can auto-recreate connection on failures (for retry errors) - it is possible to use one session per connection, but not driver/session pool - they must be shared between connections (in target state) for re-create sessions fast.

When YDB return any error best practive to destroy session and get new from pool - for next transactions ot retry the customer transaction, it mean sqlalchemy must allow transparent change connection between retries or session under the connection.

Reply:
It is not possible to manage shared driver in the SQLAlchemy along all connection, because the connection is the unit of communication with the database. I've tried to implement a custom ConnectionPool implementation (with ydb_driver inside), but the SQLAlchemy doesn't provide the connection parameters to the pool constructor.

I suggest to create ydb_driver and "small" pool for each connection + support an ability to provide an ydb_session pool managed outside the SQLAlchemy. It looks like a good compromise: 

if user wants maximum simplicity, he will just create an Engine with a connection string, SQLAlchemy will create and manage a driver and a session pool; 

if he wants maximum efficiency, he creates driver and session pool manually and provides it to the SQLAlchemy with NullPool implementation (see example in tests).

Last method is also provides an ability to use different authentication methods and to do fine tuning of session pool.